In [13]:
import itertools as it
import numpy as np
import matplotlib.pyplot as plt
import scipy.special as sp

# Exercise 1 (Sum rule, product rule)
4 red, 4 blue and 4 green marbles are in a bag. Two marbles are taken with replacement (that is the color of the ball is noticed and then the ball is put back to the bag). Find the probability that

1. First marble is blue or green,
2. Both marble are red,
3. First marble is red and the second is blue or green 
3. Both balls are of different color. (Tree diagram can help to get all possible combinations)

In [9]:

total_marbles = 12
blue_marbles = 4
green_marbles = 4
prob_blue = blue_marbles / total_marbles
prob_green = green_marbles / total_marbles
prob_blue_or_green = prob_blue + prob_green
print(f"Probability of First marble is blue or green : {prob_blue_or_green:0.03f}")


Probability of First marble is blue or green : 0.667


In [11]:
total_marbles = 12
red_marbles = 4
prob_red = red_marbles / total_marbles
prob_two_red =prob_red*prob_red
print(f"Probability of both marbleare red : {prob_two_red:0.03f}")

Probability of both marbleare red : 0.111


In [12]:
total_marbles = 12
red_marbles = 4
blue_marbles = 4
green_marbles = 4
prob_first_marble = red_marbles / total_marbles
prob_second_marble = (blue_marbles + green_marbles) / total_marbles
P= prob_first_marble*prob_second_marble
print(f"Probability of First marble is red and the second is blue or green : {P:0.03f}")

Probability of First marble is red and the second is blue or green : 0.222


(Ans. 0.667, 0.111, 0.222, 0.667 )

## Exercise 2 (Combinatorics)

Find the probability of drawing two red cards from an ordinary 52 deck of cards. (without replacement, i.e. the first card is not put back to the deck) (Ans. 0.2451)

In [22]:

total_cards = 52
red_cards = 26
total_ways = sp.comb(total_cards, 2)
red_ways = sp.comb(red_cards, 2)
prob_red_both = red_ways / total_ways
print(f"Probability of drawing two red cards in a row (without replacement):{prob_red_both:0.04f} ")


Probability of drawing two red cards in a row (without replacement):0.2451 


## Exercise 3
A card is drawn from a deck, without replacement. Find the conditional probability of $P(\text{a queen} | \text{a face card})$ 

(there are three face cards in each suit: J,Q,K) (Ans. 0.33)

In [23]:

total_cards = 52
face_cards = 3 * 4
queens = 4
prob_queen_given_face = queens / face_cards
print(f"Conditional probability of drawing a queen given that the card is a face card: {prob_queen_given_face:0.02f}" )


Conditional probability of drawing a queen given that the card is a face card: 0.33


## Exercise 4

Simulate the probability of the part 2 in exercise 1.

In [7]:
import random
bag = ['R','R','R','R','G','G','G','G','B','B','B','B']
count = 0
num_sim = 100
for i in range(num_sim):
    # draw two marbles with replacement
    draw1 = random.choice(bag)
    draw2 = random.choice(bag)
    
    # check if both marbles are red
    if draw1 == 'R' and draw2 == 'R':
        count += 1
P = count / num_sim

print(f"The probability of drawing two red marbles is: {P:0.02f}", )


The probability of drawing two red marbles is: 0.12


## Exercise 5

A local animal shelter has 24 dogs and 18 cats available for adoption. Eight of the dogs and 6 of the cats are male. Find the following conditional probabilities if an animal is selected at random.

1. The pet is a male, given that it is a cat. (Ans:0.333)
2. The pet is a cat, given that it is a female. (Ans. 0.429)
3. The pet is a female, given that it is a dog. (Ans. 0.666)

In [14]:
total_pets= 42
dogs= 24
cats= 18
dogs_male= 8
cats_male=6
P1= cats_male/cats
P2= cats/total_pets
P3= 1- dogs_male/dogs
print(f"Probabilty of The pet is a male, given that it is a cat is {P1:0.3f}")
print(f"Probabilty of The pet is cat, given that it is a female is {P2:0.3f}")
print(f"Probabilty of The pet is a female, given that it is a dog is {P3:0.3f}")


Probabilty of The pet is a male, given that it is a cat is 0.333
Probabilty of The pet is cat, given that it is a female is 0.429
Probabilty of The pet is a female, given that it is a dog is 0.667


## Exercise 6

At a college, 20% of the students take Finite Math, 30% take History, and 5% take both Finite Math and History. If a student is chosen at random, find the following conditional probabilities.

    The student is taking Finite Math given that he is taking History.
    The student is taking History assuming that he is taking Finite Math.
    
(Ans. 0.166,  0.250)


In [16]:
FM= 0.2
H= 0.3
both= 0.05 
P1= both/H
P2= both/FM
print(f"Probabilty of The student is taking Finite Math given that he is taking History is {P1:0.3f}")
print(f"Probabilty of The student is taking History assuming that he is taking Finite Math is {P2:0.3f}")

Probabilty of The student is taking Finite Math given that he is taking History is 0.167
Probabilty of The student is taking History assuming that he is taking Finite Math is 0.250


## Exercise 7

70 % of the kids who visit a doctor have a fever and 30 % of kids with a fever have sore throats. What is the probability that a kid who goes to a doctor has a fever and a sore throat? (Ans. 0.21)

In [18]:
P_fever = 0.7
P_sore_throat = 0.3
P1 = P_fever*P_sore_throat

print(f"Probabilty of a kid who goes to a doctor has a fever and a sore throat is {P1:0.2f}")

Probabilty of a kid who goes to a doctor has a fever and a sore throat is 0.21


## Exercise 8

A junk box contains twelve old batteries, five of which are totally dead. You start picking them one at a time and testing them. You do not put the tested batteries back to the junk box. Find the probabilities of the following outcomes.

1. The first two batteries are good. (Ans. 0.318)
2. At least one of the first three works. (Ans. 0.118)
3. You have to pick 5 batteries in order to find one that works. (Ans. 0.009)

In [34]:
from math import perm
num_good = 12 - 5
num_bad = 5
total_ways = perm(num_good + num_bad, 2)
good_ways = perm(num_good, 2)
P = good_ways / total_ways

print(f"Probabilty of The first two batteries are good is {P:0.3f}")

Probabilty of The first two batteries are good is 0.318


In [45]:
from math import perm
num_good = 7
num_bad = 5
total_ways = perm(num_good + num_bad, 3)
bad_ways = perm(num_bad, 3)
P1 = 1- bad_ways / total_ways

print(f"Probabilty of At least one of the first three works is {P1:0.3f}")

Probabilty of At least one of the first three works is 0.955


## Exercise 9

Company A supplies 40% of the computers sold and is late 5% of the time. Company B supplies 30% of the computers sold and is late 3% of the time. Company C supplies another 30% and is late 2.5% of the time. What is the probability that a computer arrives late? (Hint Tree diagram or total probability formula)

In [49]:

A = 0.40
late_a = 0.05
B = 0.30
late_b = 0.03
C = 0.30
late_c = 0.025

P= (A * late_a) + (B * late_b) + (C * late_c)

print(f"The probability that a computer arrives late is: {P:0.3f}" )


The probability that a computer arrives late is: 0.037


## Exercise 10

One urn has 4 red balls and 1 white ball; a second urn has 2 red balls and 3 white balls. A single card is randomly selected from a standard deck. If the card is less than 5 (aces count as 1), a ball is drawn out of the first urn; otherwise a ball is drawn out of the second urn. If the drawn ball is red, what is the probability that it came out of the second urn? (Hint: Tree diagram or Bayes)

In [2]:

Urn1 = 4/13
Urn2 = 9/13
P_R_given_Urn1 = 4/5
P_R_given_Urn2 = 2/5
P_R = P_R_given_Urn1 * Urn1 + P_R_given_Urn2 * Urn2
P_S2_given_R = P_R_given_Urn2 * Urn2 / P_R
print(f"The probability that the ball was drawn from the second urn given that a red ball was drawn is: {P_S2_given_R:0.3f}")


The probability that the ball was drawn from the second urn given that a red ball was drawn is: 0.529


## Exercise 11

A basketball team is to play two games in a tournament. The probability of winning the first game is 0.10. If the first game is won, the probability of winning the second game is 0.15. If the first game is lost, the probability of winning the second game is 0.25. What is the probability the first game was won if the second game is lost? (Hint: tree diagram or Bayes formula)

In [3]:

W1 = 0.10
L1 = 0.90
W2_given_W1 = 0.15
W2_given_L1 = 0.25
L2_given_W1 = 1 - W2_given_W1
L2_given_L1 = 1 - W2_given_L1
L2 = L2_given_W1 * W1 + L2_given_L1 * L1
W2 = W2_given_W1 * W1 + W2_given_L1 * L1
W1_given_L2 = W2_given_L1 * L1 /L2

print(f"The probability that the first game was won given that the second game was lost is: {W1_given_L2:0.3f}")


The probability that the first game was won given that the second game was lost is: 0.296


## Exercise 12
To evaluate a new test for detecting Hansen’s disease, a group of people 5% of which are known to have Hansen’s disease are tested. The test finds Hansen’s disease among 98% of those with the disease and 3% of those who don’t. What is the probability that someone testing positive for Hansen’s disease under this new test actually has it? (Bayes)

In [4]:
#  probabilty A stand for hansen disease B stand for positive test
A = 0.05
notA = 0.95
B_given_A = 0.98
B_given_notA = 0.03
B = B_given_A * A + B_given_notA * notA
A_given_B = B_given_A * A/B

print(f"The probability of having Hansen's disease given that the test is positive is: {A_given_B:0.3f}")


The probability of having Hansen's disease given that the test is positive is: 0.632
